src/components/inputs/LeaseSelectInput.js

```diff
@@ -22,27 +31,86 @@ const LeaseSelectInput = ({
   onBlur,
   onChange,
   placeholder,
-  relatedLeases,
+  leaseHistoryItems,
   value,
 }: Props) => {
-  const getLeaseOptions = (leases: Array<Object>): Array<Object> =>
-    leases
-      .filter((lease) => relatedLeases.find((relatedLease) => lease.id === relatedLease.lease.id) ? false : true)
+  const leaseAttributes = getLeaseAttributes(store.getState());
+  const getHistoryItemOptions = (leases: Array<Object>, plotSearches: Array<Object>,plotApplications: Array<Object>, areaSearches: Array<Object>): Array<Object> => {
+    const filterOutExisting = (item) => leaseHistoryItems.find((leaseHistoryItem) => item.id === leaseHistoryItem.lease.id) ? false : true
+        
+    leases = leases
+      .filter(filterOutExisting)
       .map((lease) => {
+        const stateOptions = getFieldOptions(leaseAttributes, LeaseFieldPaths.STATE)
         return {
           value: lease.id,
-          label: getContentLeaseIdentifier(lease),
+          label: `${getContentLeaseIdentifier(lease) || ''}, ${getLabelOfOption(stateOptions, lease.state) || ''}`,
+          type: 'lease'
         };
       });
 
-  const getLeases = debounce(async(inputValue: string, callback: Function) => {
+    plotSearches = plotSearches
+      .filter(filterOutExisting)
+      .map((plotSearch) => {
+        const maxLength = 24
+        const displayName = plotSearch.name.length > maxLength ? plotSearch.name.substr(0,maxLength) + "..." : plotSearch.name
+        return {
+          value: plotSearch.id,
+          label: `${displayName}, Haku`,
+          type: 'related_plot_application',
+          content_type_model: "plotsearch"
+        };
+      });
+
+    plotApplications = plotApplications
+      .filter(filterOutExisting)
+      .map((plotApplication) => {
+        return {
+          value: plotApplication.id,
+          label: `${plotApplication.application_identifier}, Hakemus`,
+          type: 'related_plot_application',
+          content_type_model: "targetstatus"
+        };
+      });
+
+    areaSearches = areaSearches
+      .filter(filterOutExisting)
+      .map((areaSearch) => {
+        return {
+          value: areaSearch.id,
+          label: `${areaSearch.identifier}, Aluehakemus`,
+          type: 'related_plot_application',
+          content_type_model: "areasearch"
+        };
+      });
+
+      return [...leases, ...plotSearches, ...plotApplications, ...areaSearches].sort((a, b) => (a.label && b.label) && a.label > b.label ? 1 : -1).slice(0,10)
+  }
+
+  const getHistoryItems = debounce(async(inputValue: string, callback: Function) => {
     const leases = await fetchLeases({
       succinct: true,
       identifier: inputValue,
       limit: 10,
     });
+    
+
+    const plotSearches = await fetchPlotSearches({
```

Comment:
How about using Promise.all() here?

Reply:
Would like to receive insights on this question, why not implement with Promise.all?

Reply:
In brief, it would lead to unnecessary complications in the code. As far as I understand the logic, Promise.all() would create a single list of results that would then passed into the getHistoryItemOptions callback function. But if we do that, we would need to sort out the types of the four different kinds of data anyway in order to get the correct values for the label, type and content_type_model. It is easier to sort these out when the result sets from each endpoint is kept separate and then later combined into a single list.

Reply:
Promise.all() would not create a single list of result, it would return each Promise individually. The benefit would be that the fetches are run simultaneously and would possibly take less time. How I see the current code is that it does one fetch, waits for response, does another fetch, waits, and so on.

I suggested it purely to make all the calls at once and reduce the time waiting for the responses one by one.

Reply:
Done